qualtran/bloqs/arithmetic/addition.py

```diff
@@ -266,6 +268,120 @@ def _add_oop_large() -> OutOfPlaceAdder:
+        k: The classical integer value to be added to x.
+        cvs: A tuple of control variable settings. Each entry specifies whether that
+            control line is a "positive" control (`cv[i]=1`) or a "negative" control `0`.
+        signed: A boolean condition which controls whether the x register holds a value represented
+            in 2's Complement or Unsigned. This affects the ability to add a negative constant.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+
+    References:
+        [Improved quantum circuits for elliptic curve discrete logarithms](https://arxiv.org/abs/2001.09580) Fig 2a
+    """
+
+    bitsize: int
+    k: int
+    cvs: Tuple[int, ...] = field(converter=lambda v: (v,) if isinstance(v, int) else tuple(v))
+    signed: bool
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        if len(self.cvs) > 0:
+            return Signature(
+                [Register('ctrl', bitsize=len(self.cvs)), Register('x', bitsize=self.bitsize)]
+            )
+        else:
+            return Signature([Register('x', bitsize=self.bitsize)])
+
+    def apply(self, *args) -> Union[int, Iterable[int]]:
+        target_val = args[-1]
+        new_target_val = target_val + self.k
+
+        if self.cvs and args[0] != int(''.join(str(x) for x in self.cvs), 2):
+            new_target_val = target_val
+        ret = (args[0], new_target_val) if self.cvs else (new_target_val,)
+        return ret
+
+    def on_classical_vals(self, *args) -> Dict[str, 'ClassicalValT']:
+        return dict(zip([reg.name for reg in self.signature], self.apply(*args)))
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **regs: SoquetT) -> Dict[str, 'SoquetT']:
+
+        # Assign registers to variables and allocate ancilla bits for classical integer k.
+        x = regs['x']
+        if len(self.cvs) > 0:
+            ctrls = regs['ctrl']
+        else:
+            ctrls = None
+        k = bb.allocate(n=self.bitsize)
+
+        # Get binary representation of k and split k into separate wires.
+        k_split = bb.split(k)
+        if self.signed:
+            binary_rep = list(iter_bits_twos_complement(self.k, self.bitsize))
+        else:
+            binary_rep = list(iter_bits(self.k, self.bitsize))
+        binary_rep.reverse()
+
+        # Apply XGates to qubits in k where the bitstring has value 1. Apply CNOTs when the gate is
+        # controlled.
+        for i in range(self.bitsize):
+            if binary_rep[i] == 1:
+                if len(self.cvs) > 0:
+                    ctrls, k_split[i] = bb.add(
+                        MultiControlPauli(cvs=self.cvs, target_gate=XGate()),
+                        controls=ctrls,
+                        target=k_split[i],
+                    )
+                else:
+                    k_split[i] = bb.add(XGate(), q=k_split[i])
+
+        # Rejoin the qubits representing k for in-place addition.
+        k = bb.join(k_split)
```

Comment:
can you open an issue to track turning this into `IntState`? It's blocked on being able to control it, right?

Reply:
Will do, but for my own knowledge what is IntState? It has something to do with the controlled bloq protocol, right?

Reply:
`qualtran.bloqs.basic_gates.z_basis.IntState` is a bloq that represents allocating a state initialized to an unsigned integer